lib/benchmarks/src/Cardano/Wallet/Benchmarks/Collect.hs

```diff
@@ -0,0 +1,461 @@
+
+instance ToField Semantic where
+    toField = toField . T.unwords . reverse . unMkSemantic
+
+-- | Create a new semantic from a list of words.
+-- It will frop empty words.
+-- It will replace spaces with dashes.
+mkSemantic :: [Text] -> Semantic
+mkSemantic = Semantic . filter (not . T.null) . fmap (T.replace " " "-")
+
+unMkSemantic :: Semantic -> [Text]
+unMkSemantic (Semantic xs) = toList xs
+
+-- | Units for a result.
+data Units
+    = Seconds
+    | Milliseconds
+    | Microseconds
+    | Nanoseconds
+    | Bytes
+    | KiloBytes
+    | MegaBytes
+    | GigaBytes
+    | Count
+
+instance Show Units where
+    show = showUnits
+
+instance FromField Units where
+    parseField "s" = pure Seconds
+    parseField "ms" = pure Milliseconds
+    parseField "us" = pure Microseconds
+    parseField "ns" = pure Nanoseconds
+    parseField "B" = pure Bytes
+    parseField "KB" = pure KiloBytes
+    parseField "MB" = pure MegaBytes
+    parseField "GB" = pure GigaBytes
+    parseField "count" = pure Count
+    parseField _ = fail "Invalid units"
+
+instance ToField Units where
+    toField = B8.pack . showUnits
+
+showUnits :: Units -> String
+showUnits Seconds = "s"
+showUnits Milliseconds = "ms"
+showUnits Microseconds = "us"
+showUnits Nanoseconds = "ns"
+showUnits Bytes = "B"
+showUnits KiloBytes = "KB"
+showUnits MegaBytes = "MB"
+showUnits GigaBytes = "GB"
+showUnits Count = "count"
+
+-- | A result with a value and units.
+data Result = Result
+    { resultValue :: Double
+    , resultUnits :: Units
+    , resultIterations :: Int
+    }
+
+instance Show Result where
+    show (Result value units iterations) =
+        show value
+            ++ " "
+            ++ show units
+            ++ " ("
+            ++ show iterations
+            ++ " iterations)"
+
+-- | A benchmark result with a semantic and a value.
+data Benchmark = Benchmark
+    { benchmarkSemantic :: Semantic
+    , benchmarkResult :: Result
+    }
+    deriving (Show)
+
+-- | A class to collect 'Benchmark' from values of any type.
+class HasResults m r where
+    resultsOf :: r -> m [Benchmark]
+
+instance Monad m => HasResults m Benchmark where
+    resultsOf = pure . pure
+
+instance HasSeverityAnnotation Benchmark where
+    getSeverityAnnotation _ = Notice
+
+instance ToText Benchmark where
+    toText (Benchmark sem res) =
+        T.unwords
+            [ T.unwords $ reverse $ unMkSemantic sem
+            , T.pack $ show res
+            ]
+
+instance ToNamedRecord Benchmark where
+    toNamedRecord (Benchmark semantic (Result value units iterations)) =
+        namedRecord
+            [ "semantic" .= toField semantic
+            , "value" .= toField value
+            , "units" .= toField units
+            , "iterations" .= toField iterations
+            ]
+
+changeBenchmarkSemantic :: (Semantic -> Semantic) -> Benchmark -> Benchmark
+changeBenchmarkSemantic f (Benchmark sem res) = Benchmark (f sem) res
+
```

Comment:
what about documenting the job here - what is typical input here (ie. header+row) and how it is translated into report?

Reply:
could be in the form of unit test for toNamedRecord and parseNamedRecord or just as comment